aggregators/aggregator.go

```diff
@@ -163,277 +90,168 @@ func New(opts ...Option) (*Aggregator, error) {
+	case <-a.closed:
 	default:
+		close(a.closed)
 	}
 
 	var errs []error
-	var totalBytesIn int64
-	cmk := CombinedMetricsKey{ID: id}
-	for _, ivl := range a.cfg.AggregationIntervals {
-		cmk.ProcessingTime = a.processingTime.Truncate(ivl)
-		cmk.Interval = ivl
-		for _, e := range *b {
-			bytesIn, err := a.aggregateAPMEvent(ctx, cmk, e)
-			if err != nil {
-				errs = append(errs, err)
-			}
-			totalBytesIn += int64(bytesIn)
+	a.writers.Range(func(_, value any) bool {
+		writer := value.(*Writer)
+		if err := writer.Close(); err != nil {
+			errs = append(errs, fmt.Errorf("failed to close writer: %w", err))
+		}
+		return true
+	})
+	if a.harvestingStopped != nil {
+		select {
+		case <-ctx.Done():
+			errs = append(errs, fmt.Errorf("context cancelled while waiting for harvesting to stop: %w", ctx.Err()))
+		case <-a.harvestingStopped:
 		}
-		a.cachedEvents.add(ivl, id, float64(len(*b)))
 	}
-
-	cmIDAttrSet := attribute.NewSet(cmIDAttrs...)
-	a.metrics.RequestsTotal.Add(ctx, 1, metric.WithAttributeSet(cmIDAttrSet))
-	a.metrics.BytesIngested.Add(ctx, totalBytesIn, metric.WithAttributeSet(cmIDAttrSet))
-	if len(errs) > 0 {
-		a.metrics.RequestsFailed.Add(ctx, 1, metric.WithAttributeSet(cmIDAttrSet))
-		return fmt.Errorf("failed batch aggregation:\n%w", errors.Join(errs...))
+	if a.db != nil {
+		if err := a.finalHarvest(ctx); err != nil {
+			errs = append(errs, fmt.Errorf("failed to perform final harvest: %w", err))
+		}
+		if err := a.db.Close(); err != nil {
+			errs = append(errs, fmt.Errorf("failed to close pebble: %w", err))
+		}
+		a.db = nil
 	}
-	return nil
+	if err := a.metrics.CleanUp(); err != nil {
+		errs = append(errs, fmt.Errorf("failed to cleanup instrumentation: %w", err))
+	}
+	return errors.Join(errs...)
 }
 
-// AggregateCombinedMetrics aggregates partial metrics into a bigger aggregate.
-// This function will return an error if the aggregator's Run loop has errored
-// or has been explicitly stopped. However, it doesn't require aggregator to be
-// running to perform aggregation.
-func (a *Aggregator) AggregateCombinedMetrics(
-	ctx context.Context,
-	cmk CombinedMetricsKey,
-	cm *aggregationpb.CombinedMetrics,
-) error {
-	cmIDAttrs := a.cfg.CombinedMetricsIDToKVs(cmk.ID)
-	traceAttrs := append(append([]attribute.KeyValue{}, cmIDAttrs...),
-		attribute.String(aggregationIvlKey, formatDuration(cmk.Interval)),
-		attribute.String("processing_time", cmk.ProcessingTime.String()))
-	ctx, span := a.cfg.Tracer.Start(ctx, "AggregateCombinedMetrics", trace.WithAttributes(traceAttrs...))
-	defer span.End()
-
+// NewWriter returns a new Writer, for adding metric values to be aggregated.
+//
+// The returned Writer will be associated with the aggregator, and will be
+// automatically closed (with any buffered writes committed) when the aggregator
+// is closed.
+//
+// NewWriter will return an error if the aggregator has already been stopped.
+func (a *Aggregator) NewWriter() (*Writer, error) {
 	a.mu.Lock()
```

Comment:
`RLock` should be sufficient but it leads to worse lock contention and worse performance in other places.

Reply:
That's very surprising, given that NewWriter should not be in any hot code paths...